docs/architecture/devops-for-aspnet-developers/actions-vs-pipelines.md

```diff
@@ -0,0 +1,121 @@
+
+## Agents and runners
+
+Before you examine pipelines themselves, you should consider how these pipelines _execute_. Both GitHub Actions and Azure Pipelines are really _orchestration engines_: when a pipeline is triggered, the system finds an "agent" and tells the agent to execute the jobs defined in the pipeline file.
+
+Azure Pipelines run on _agents_. The agent is written in .NET, so it will run wherever .NET can run: Windows, Mac and Linux. Agents can even run in containers! Agents are registered to a [pool](https://docs.microsoft.com/azure/devops/pipelines/agents/pools-queues?view=azure-devops) in Azure Pipelines or to a repository or organization in GitHub. Agents can be _hosted_ or _private_.
+
+GitHub Workflows execute on _runners_. The runner code is essentially a fork of the Azure Pipelines code, so it is very similar. It is also cross-platform and you can also utilize _hosted_ or _self-hosted_ runners.
+
+### Hosted agents and runners
+
+Hosted agents (Azure Pipelines) and hosted runners (GitHub) are agents that are spun up and managed by Azure DevOps or GitHub respectively. You do not need to maintain any build infrastructure. When a pipeline triggers that targets a hosted agent, an instance of the specified agent image is spun up. The job is run by the agent on the instance, and once the job completes, the instance is destroyed. The same applies for hosted runners running GitHub workflows.
+
+> [!INFORMATION]
+> The list of software installed on Azure Pipeline images is listed in [this repo](https://github.com/actions/virtual-environments/tree/master/images). You can click into the platform folder and examine the README.md files. You can find information on GitHub hosted runners [here](https://docs.github.com/en/actions/reference/specifications-for-github-hosted-runners).
+
+### Private agents and self-hosted runners
+
+There are times when you can't utilize hosted images:
+
+- you require SDKs or other software that is not installed on the images
+- you need to access resources that are not public (such as an internal Sonarqube server or an internal Artifactory instance)
+- you need to deploy to private networks
+- you need to install licenses for 3rd party software required for building your code
+- you need more storage or memory than is provided to the hosted agent images
+- you need more time than the maximum build time limit for hosted agents
+
+> [!INFORMATION]
+> It is possible to install tools and SDKs when running pipelines on hosted agents. If the install steps do not take long, this is viable. However, if the tools/software take a long time to install, then you may be better off with a private agent or self-hosted runner, since the install steps will need to execute for every run of the workflow.
```

Comment:
```suggestion
> It's possible to install tools and SDKs when running pipelines on hosted agents. If the install steps don't take long, this is viable. However, if the tools/software take a long time to install, you may be better off with a private agent or self-hosted runner, since the install steps will need to execute for every run of the workflow.
```